Update "empty" web project templates to use minimal hostinng and routing

halter73 opened the pull request:
## Description

Update the "Empty" project templates (used by `dotnet new empty`/`dotnet new web` and VS) to use the new "minimal hosting" APIs that were just merged minutes ago for preview4 by #31825.

## Customer Impact

We're looking to collect feedback on our new "minimal hosting" and "minimal routing" API which we will present at BUILD. This updates the "empty" web project templates to make it easier to customer trying preview4 after BUILD to test out these new APIs.

## Regression?
- [ ] Yes
- [x] No

The template behaves as it did before. It does not include all the `usings`/`opens` for unused namespace. That might change based on code review feedback.

## Risk
- [ ] High
- [x] Medium
- [ ] Low

This is a pretty big change to the empty web project template. It might confuse people. It is *just* a template change though.

@glennc looked up some usage statistics and this is one of the less popular than the MVC and API templates.

## Verification
- [x] Manual (required)
- [x] Automated

## Packaging changes reviewed?
- [ ] Yes
- [ ] No
- [x] N/A


Addresses #30665

halter73 committed:
Minimize EmptyWeb-CSharp project template

halter73 committed:
Add UseDeveloperExceptionPage() back

halter73 committed:
Minimize EmptyWeb-FSharp project template

halter73 committed:
Add MinimalSampleFSharp

cartermp commented on src/Http/samples/MinimalSampleFSharp/Program.fs:
```diff
@@ -0,0 +1,20 @@
+    if app.Environment.IsDevelopment() then
+        app.UseDeveloperExceptionPage() |> ignore
+
+    app.MapGet("/", Func<string>(fun () -> "Hello World!")) |> ignore
```
Is the explicit `Func` construction needed here? Might be able to elide that

halter73 replied:
It seems like it. Without it, I see the following error:

```
C:\dev\dotnet\aspnetcore\src\Http\samples\MinimalSampleFSharp\Program.fs(12,34): error FS0001: This expression was expected to have type↔    'AspNetCore.Http.HttpContext'    ↔but here has type↔    'unit' [C:\dev\dotnet\aspnetcore\src\Http\samples\MinimalSampleFSharp\MinimalSampleFSharp.fsproj]
```

davidfowl replied:
This is the F# language feature we want

cartermp commented on src/Http/samples/MinimalSampleFSharp/Program.fs:
```diff
@@ -0,0 +1,20 @@
+    app.MapGet("/", Func<string>(fun () -> "Hello World!")) |> ignore
+
+    app.MapGet("/plaintext", Func<string>(fun () -> "Hello, World!")) |> ignore
+    app.MapGet("/json", Func<obj>(fun () -> {| message = "Hello, World!" |} :> obj)) |> ignore
```
Hmm, weird that the cast to `obj` is needed here. F# anonymous records compile down to anonymous objects for interop

cartermp replied:
If it's needed, then this might be cleaner:

`upcast {| message = "Hello, World!" |} `

cartermp replied:
You can also try `Func<_>` and see if no cast is necessary

halter73 commented on src/ProjectTemplates/Web.ProjectTemplates/content/EmptyWeb-CSharp/Program.cs:
```diff
@@ -1,26 +1,14 @@
+    app.UseDeveloperExceptionPage();
 }
+
+app.MapGet("/", (Func<string>)(() => "Hello World!"));
```
I don't like that we still have to explicitly cast to `Func<string>` until we can consume the changes in https://github.com/dotnet/roslyn/pull/52448, but I still think this is already a big improvement overall.

davidfowl replied:
@jaredpar @cston

jaredpar replied:
The bits should be lining up for preview 4 to allow this cast to be removed.

halter73 commented on src/ProjectTemplates/Web.ProjectTemplates/content/EmptyWeb-CSharp/Program.cs:
```diff
@@ -1,26 +1,14 @@
-using Microsoft.Extensions.Configuration;
+using Microsoft.AspNetCore.Builder;
 using Microsoft.Extensions.Hosting;
-using Microsoft.Extensions.Logging;
```
Do we care about keeping the extra usings?

LadyNaggaga replied:
Not really.  Logging fits in this scenario so, it isn't entirely unnecessary.

halter73 committed:
upcast

Tratcher commented on src/ProjectTemplates/Web.ProjectTemplates/content/EmptyWeb-CSharp/Program.cs:
```diff
@@ -1,26 +1,14 @@
+
+app.MapGet("/", (Func<string>)(() => "Hello World!"));
+
+await app.RunAsync();
```
Program+Startup before: 66 lines
after: 14 lines. 

It's a lot emptier than it used to be 😁.

davidfowl reviewed (approved):
Make the "empty" project, "really empty"

halter73 committed:
Fix tests

MCCshreyas commented:
Go..Go..Go SingleFile 😅😅

jkotalik commented:
This plus making Kestrel the default server brings a smile to my face 😄. True modernization.

Tratcher commented on src/Http/samples/MinimalSample/Program.cs:
```diff
@@ -1,10 +1,15 @@
+using Microsoft.Extensions.Hosting;
 
-await using var app = WebApplication.Create();
+await using var app = WebApplication.Create(args);
```
I'm just now appreciating that this removes any ability to configure DI, options, etc., those require the builder. What's the simplest version of this that makes the builder discoverable?
```C#
var builder = WebApplication.CreateBuilder(args);
await using var app = builder.Build();
```
That extra line seems worth it.

halter73 replied:
This is just the sample. I'm guessing you want to change this in the template as well. @LadyNaggaga thoughts?

HaoK replied:
Can people actually do anything interesting without configuring DI/options?

Tratcher replied:
With WebApplicaiton they can do middleware, routes, and IConfiguration, but not set up logging, options, DI, servers, etc.

HaoK replied:
Then we just need a magic autobind so you can just set options by convention in config!...kidding

davidfowl replied:
I dunno, maybe. We basically need to go against our natural instincts here 😄 

cc @LadyNaggaga

halter73 replied:
We're adding the builder.

halter73 committed:
Show WebApplicationBuilder

halter73 committed:
_

Commit:
Fixes #32028: Investigate Template Baseline tests errors

Latest `dotnet new` doesn't work with relative path for `--debug:custom-hive`... Hence make it absolute
Also switched to using `--debug:disable-sdk-templates` to avoid potential conflict between SDK and tests installed templates.

halter73 commented:
/azp run

halter73 committed:
Update SDK to 6.0.100-preview.4.21222.6

halter73 commented on global.json:
```diff
@@ -1,9 +1,9 @@
 {
   "sdk": {
-    "version": "6.0.100-preview.4.21216.8"
+    "version": "6.0.100-preview.4.21222.6"
```
Heads up @dotnet/aspnet-build. I'm updating the preview4 branch SDK version so we can consume the `WebApplication` related types from our new "empty" web project templtates.

dougbu replied:
If these new types come from dotnet/aspnetcore, something is very wrong if they aren't just showing up. Fine if, on the other hand, the main thing we need is @DavidKarlas' dotnet/templating#3078 fix.

halter73 replied:
I needed to update this to get EmptyWebTemplateTest.EmptyWebTemplateCSharp EmptyWebTemplateTest.EmptyWebTemplateFSharp passing locally and on helix, so there's definitely room for improvement here. Not sure how common it is to update the templates to use very-newly-added APIs.

HaoK replied:
My understanding of the fix to templating is that the regression broke the custom hive logic we were using to make our template tests find our updated templates, so that's why stefan had to update the sdk.  The templating workaround PR should also fix the tests I would assume

Pilchie commented:
:eyes:

dougbu reviewed (approved):
Looks fine from a build perspective
